pkg/ingester/recovery.go

```diff
@@ -0,0 +1,395 @@
+	inst := r.ing.getOrCreateInstance(series.UserID)
+
+	// TODO(owen-d): create another fn to avoid unnecessary label type conversions.
+	stream, err := inst.getOrCreateStream(logproto.Stream{
+		Labels: client.FromLabelAdaptersToLabels(series.Labels).String(),
+	}, true, nil)
+
+	if err != nil {
+		return err
+	}
+
+	added, err := stream.setChunks(series.Chunks)
+	if err != nil {
+		return err
+	}
+	r.ing.metrics.recoveredChunksTotal.Add(float64(len(series.Chunks)))
+	r.ing.metrics.recoveredEntriesTotal.Add(float64(added))
+
+	// now store the stream in the recovery map under the fingerprint originally recorded
+	// as it's possible the newly mapped fingerprint is different. This is because the WAL records
+	// will use this original reference.
+	got, _ := r.users.LoadOrStore(series.UserID, &sync.Map{})
+	streamsMap := got.(*sync.Map)
+	streamsMap.Store(series.Fingerprint, stream)
+	return nil
+}
+
+// SetStream is responsible for setting the key path for userIDs -> fingerprints -> streams.
+// Internally, this uses nested sync.Maps due to their performance benefits for sets that only grow.
+// Using these also allows us to bypass the ingester -> instance -> stream hierarchy internally, which
+// may yield some performance gains, but is essential for the following:
+// Due to the use of the instance's fingerprint mapper, stream fingerprints are NOT necessarily
+// deterministic. The WAL uses the post-mapped fingerprint on the ingester that originally
+// created the stream and we ensure that said fingerprint maps correctly to the newly
+// created stream during WAL replay, even if the new in memory stream was assigned a different
+// fingerprint from the mapper. This is paramount because subsequent WAL records will use
+// the fingerprint reported in the WAL record, not the potentially differing one assigned during
+// stream creation.
+func (r *ingesterRecoverer) SetStream(userID string, series record.RefSeries) error {
+	inst := r.ing.getOrCreateInstance(userID)
+
+	stream, err := inst.getOrCreateStream(
+		logproto.Stream{
+			Labels: series.Labels.String(),
+		},
+		true,
+		nil,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Now that we have the stream, ensure that the userID -> fingerprint -> stream
+	// path is set properly.
+	got, _ := r.users.LoadOrStore(userID, &sync.Map{})
+	streamsMap := got.(*sync.Map)
+	streamsMap.Store(series.Ref, stream)
+	return nil
+}
+
+func (r *ingesterRecoverer) Push(userID string, entries RefEntries) error {
+	out, ok := r.users.Load(userID)
+	if !ok {
+		return errors.Errorf("user (%s) not set during WAL replay", userID)
+	}
+
+	s, ok := out.(*sync.Map).Load(entries.Ref)
+	if !ok {
+		return errors.Errorf("stream (%d) not set during WAL replay for user (%s)", entries.Ref, userID)
+	}
+
+	// ignore out of order errors here (it's possible for a checkpoint to already have data from the wal segments)
+	_ = s.(*stream).Push(context.Background(), entries.Entries, nil)
```

Comment:
you could log/warn though in case it's no out of order, I don't like swallowing.

Reply:
I could, but it's expected for this to be hit very often. Since we checkpoint series across a duration, the series checkpointed at the end will likely include most of the entries that the WAL segments have, meaning that during replay, the segments will try to push a ton of data that already exists via recovered checkpoints. This results in a ton of out of orders while the segments catch up to the checkpoint. It's expected behavior, but I don't think we should log. Perhaps I can add a metric here though.

Reply:
I understand, it not easy my point was that err could be something else. Although I checked and it's always out of order, at least for now.